File: Data Structures & Algorithms/surrounded-regions/submission-0.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        rows,cols = len(board), len(board[0])
        directions = [(-1,0),(1,0),(0,-1),(0,1)]
        visited = [[False] * cols for _ in range(rows)]
        queue = deque()

        # add neighbor cells
        for c in range(cols):
            if board[0][c] == "O":
                queue.append((0,c))
                board[0][c] = "S"
            if board[rows-1][c] == "O":
                queue.append((rows-1,c))
                board[rows-1][c] = "S"
        for r in range(rows):
            if board[r][0] == "O":
                queue.append((r,0))
                board[r][0] = "S"
            if board[r][cols-1] == "O":
                queue.append((r,cols-1))
                board[r][cols-1] = "S"

        # bfs - change neighbors of Os to safe
        while queue:
            r,c = queue.popleft()

            for dr,dc in directions:
                nr,nc = r+dr,c+dc

                if 0<=nr<rows and 0<=nc<cols:
                    if board[nr][nc] == "O":
                        board[nr][nc] = "S"
                        queue.append((nr, nc))
        
        for r in range(rows):
            for c in range(cols):
                if board[r][c] == "O":
                    board[r][c] = "X"
        for r in range(rows):
            for c in range(cols):
                if board[r][c] == "S":
                    board[r][c] = "O"

        

File: Data Structures & Algorithms/valid-sudoku/submission-0.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        unordered_map<int, vector<int>> rows;
        unordered_map<int, vector<int>> cols;
        unordered_map<int, vector<int>> squares;
        
        
        for (int i = 0; i< 9; i++) {
            for (int j=0; j<9; j++) {
                int sqI = i/3 *3 +j/3;
                if (board[i][j] != '.') {
                    if (find(rows[i].begin(), rows[i].end(), board[i][j]) != rows[i].end() ||
                find(cols[j].begin(), cols[j].end(), board[i][j]) != cols[j].end() ||
               find(squares[sqI].begin(), squares[sqI].end(), board[i][j]) != squares[sqI].end()) 
                return false;
                rows[i].push_back(board[i][j]);
                cols[j].push_back(board[i][j]);
                squares[sqI].push_back(board[i][j]);
                }
                
            }
        }
        return true;
    }
};
